File: backend/Ronboard.Api/Hubs/SessionHub.cs
namespace Ronboard.Api.Hubs;

using Microsoft.AspNetCore.SignalR;
using Ronboard.Api.Services;

public class SessionHub(SessionManager sessionManager) : Hub
{
    public async Task SendInput(Guid sessionId, string data)
    {
        await sessionManager.SendInputAsync(sessionId, data);
    }

    public async Task SendMessage(Guid sessionId, string message)
    {
        await sessionManager.SendStreamMessageAsync(sessionId, message);

        // Broadcast user message to other clients in the group (sender already has it locally)
        var msgs = sessionManager.GetStreamHistory(sessionId);
        var userMsg = msgs.LastOrDefault(m => m.Type == "user_message");
        if (userMsg is not null)
            await Clients.OthersInGroup(sessionId.ToString()).SendAsync("StreamMessage", sessionId, userMsg);
    }

    public async Task JoinSession(Guid sessionId)
    {
        await Groups.AddToGroupAsync(Context.ConnectionId, sessionId.ToString());

        var session = sessionManager.Get(sessionId);
        if (session is null) return;

        if (session.Mode == Models.SessionMode.Terminal)
        {
            var history = sessionManager.GetTerminalHistory(sessionId);
            if (!string.IsNullOrEmpty(history))
                await Clients.Caller.SendAsync("TerminalHistory", history);
        }
        else
        {
            var messages = sessionManager.GetStreamHistory(sessionId);
            if (messages.Count > 0)
                await Clients.Caller.SendAsync("StreamHistory", messages);
        }
    }

    public async Task LeaveSession(Guid sessionId)
    {
        await Groups.RemoveFromGroupAsync(Context.ConnectionId, sessionId.ToString());
    }
}

File: backend/Ronboard.Api/SessionHub.fs
namespace Ronboard.Api.Hubs

open System
open Microsoft.AspNetCore.SignalR
open Microsoft.Extensions.Logging
open Ronboard.Api.Models
open Ronboard.Api.Services

type SessionHub(sessionOps: ISessionOperations, logger: ILogger<SessionHub>) =
    inherit Hub()

    member this.SendInput(sessionId: Guid, data: string) =
        task {
            logger.LogInformation(
                "User terminal input for session {SessionId} ({Length} chars)",
                sessionId,
                data.Length
            )

            do! sessionOps.SendInputAsync(sessionId, data)
        }

    member this.SendMessage(sessionId: Guid, message: string) =
        task {
            logger.LogInformation("User message for session {SessionId}: {Message}", sessionId, message)
            do! sessionOps.SendStreamMessageAsync(sessionId, message)

            let msgs = sessionOps.GetStreamHistory(sessionId)
            let userMsg = msgs |> List.tryFindBack (fun m -> m.Type = "user_message")

            match userMsg with
            | Some msg ->
                do!
                    this.Clients
                        .OthersInGroup(sessionId.ToString())
                        .SendCoreAsync("StreamMessage", [| sessionId; msg |])
            | None -> ()
        }

    member this.JoinSession(sessionId: Guid) =
        task {
            do! this.Groups.AddToGroupAsync(this.Context.ConnectionId, sessionId.ToString())

            match sessionOps.Get(sessionId) with
            | None -> ()
            | Some session ->
                match session.Mode with
                | Terminal ->
                    let history = sessionOps.GetTerminalHistory(sessionId)

                    if not (String.IsNullOrEmpty history) then
                        do! this.Clients.Caller.SendCoreAsync("TerminalHistory", [| history |])
                | Stream ->
                    let messages = sessionOps.GetStreamHistory(sessionId)

                    if messages.Length > 0 then
                        do! this.Clients.Caller.SendCoreAsync("StreamHistory", [| messages |])

                let activity = sessionOps.GetActivityState(sessionId)

                let stateName =
                    match activity with
                    | ActivityState.ToolUse -> "tool_use"
                    | ActivityState.Busy -> "busy"
                    | ActivityState.Idle -> "idle"

                do! this.Clients.Caller.SendCoreAsync("SessionActivity", [| sessionId; stateName |])
        }

    member this.LeaveSession(sessionId: Guid) =
        task { do! this.Groups.RemoveFromGroupAsync(this.Context.ConnectionId, sessionId.ToString()) }
